### ISSUE TITLE
React Native Integration Issues

### ISSUE THREAD

**Issue by Keysamm** [kettanaito/msw#1926]
Hi guys,

I have been trying to integrate **Mock Service Worker** into my application for several days. Next, I will describe the problems that arose, which I was able to fix and at which i have stuck.

First of all I followed setup provided here: https://mswjs.io/docs/integrations/react-native 

**_index.js:_**

```
// imports
async function enableMocking() {
  if (!__DEV__) {
    return;
  }

  await import('./msw.polyfills');
  const {server} = await import('./src/mocks/server');
  server.listen();
}

enableMocking().then(() => {
  AppRegistry.registerComponent(appName, () => App);
});
```

**_server.js_**

```
import {http, HttpResponse} from 'msw';
import {setupServer} from 'msw/native';

export const server = setupServer(
  http.post('http://localhost:3000/api/addData', () => {
    console.log('[ SERVER INTERCEPTOR ]');
    return HttpResponse.json({
      lastName: 'Steve',
      firstName: 'John',
    });
  }),
  http.get('http://localhost:3000/api/data', () => {
    return HttpResponse.json({
      id: '15d42a4d-1948-4de4-ba78-b8a893feaf45',
      firstName: 'John',
    });
  }),
);
```

Im providing errors from my console, as You can see its related to **_Class private methods_** , and missing some files with following path: _'/Users/distiller/react-native/packages/react-native/sdks/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js'_

<img width="854" alt="image" src="https://github.com/mswjs/msw/assets/62395388/0381feb2-561b-4e74-93da-675c8a79de9b">

After long research i changed my _index.js_ and _babel.config.js_ with following content: 

**index.js**

```
import {App} from '@app';
import {AppRegistry, Text, TextInput} from 'react-native';
import {name as appName} from './app.json';

if (__DEV__) {
  require('./msw.polyfills');

  const {server} = require('./src/mocks/server');
  server.listen();
}
AppRegistry.registerComponent(appName, () => App);
```

**babel.config.js**

```
plugins: [
      '@babel/plugin-transform-runtime',
      ['@babel/plugin-proposal-decorators', {legacy: true}],
      '@babel/plugin-transform-flow-strip-types',  // added this line
      ['@babel/plugin-transform-private-methods', {loose: true}],  // added this line
    ],
```

After this a new error occurred : 

<img width="849" alt="image" src="https://github.com/mswjs/msw/assets/62395388/f37f6750-7ef1-4dfd-96c0-cae5e62d556e">

For fixing it I created new **globals.js** file in my root app directory with following content:
`global.TextEncoder = require('text-encoding').TextEncoder;`

and imported it on top of my index.js file: 

```
import {App} from '@app';
import {AppRegistry, Text, TextInput} from 'react-native';
import {name as appName} from './app.json';
import './globals.js';
```

After all this steps my app launched successfully! Buuuuuut, as I understood requests are intercepted, but i cant get a mocked response.

Here is the code which calls **http://localhost:3000/api/addData**: 

```
const body = {
        title: Title,
        body: Body,
        is_done: false,
      };

      const res = await fetch('http://localhost:3000/api/addData', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      });
      
      const completeResponse = await res.json();
      console.log(completeResponse);
```

and here is the mocked response Im expecting: 

```
import {http, HttpResponse} from 'msw';
import {setupServer} from 'msw/native';

export const server = setupServer(
  http.post('http://localhost:3000/api/addData', () => {
    console.log('[ SERVER INTERCEPTOR ]');
    return HttpResponse.json({
      lastName: 'Steve',
      firstName: 'John',
    });
  }),
);
```

But here is my logs: 
<img width="849" alt="image" src="https://github.com/mswjs/msw/assets/62395388/6a402e30-b4f3-4a8c-b9e5-bf50be71e65f">

I tried to modify the code as follows: 

```
      const res = await fetch('http://localhost:3000/api/addData', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(body),
      });

      // const completeResponse = await res.json(); // commented this line with .json()
      console.log(res);
```

And here is the logs after console.log(res) executed:
<img width="853" alt="image" src="https://github.com/mswjs/msw/assets/62395388/c5ca1cda-1b04-4d99-a4f8-1978430d2407">

Also if Im trying to perform requests via **axios**, but also having a couple of issues:
```
 const body = {
        title: Title,
        body: Body,
        is_done: false,
      };
      const res = await axios.post('http://localhost:3000/api/addData', body);
      console.log(res);
```
<img width="856" alt="Screenshot 2023-12-13 at 13 50 16" src="https://github.com/mswjs/msw/assets/62395388/b3e61636-b503-4efb-afd4-1abf1af6c687">

Maybe I'm doing something wrong? I would be glad to hear any thoughts you have that might help me

FYI: 
**react-native : 0.73.0,
msw : 2.0.11"**

**Comment by Keysamm** [comment]
Im able to set **status**, **statusText** and receive it from MSW, but i cannot understand why my body disappear.

Looks like due to body disappearing im getting error when trying to call `.json` method on mocked response

Im providing screenshot where you can see that MSW creates **HttpResponse** instance with mocked data, but when im receiving it in my app, data disappears

<img width="752" alt="image" src="https://github.com/mswjs/msw/assets/62395388/72bccdd8-f406-4e91-a217-c27743a88ad1">

**Comment by willdawsonme** [comment]
I've run into similar issues with react native. I've had success patching msw and @mswjs/interceptors with patch-package:
- I patched `msw/native` to import the FetchInterceptor instead of XMLHttpRequestInterceptor (we use the global fetch object for our requests in react-native)
- When using FetchInterceptor, I found that response.body was undefined [on this line](https://github.com/mswjs/interceptors/blob/122a6533ce57d551dc3b59b3bb43a39026989b70/src/interceptors/fetch/index.ts#L135), so that may explain why you're having issues as well.
- I patched the line linked above to be `const response = mockedResponse.clone();` which resolved response.body being undefined.
- I removed any `"browser": null` lines in package.json.
- I've found that @msw/interceptors expects request.signal to always be defined, but that isn't the case in react-native. I've also patched [this line](https://github.com/mswjs/interceptors/blob/122a6533ce57d551dc3b59b3bb43a39026989b70/src/interceptors/fetch/index.ts#L62) to defensively add the event listener.

**Comment by anujsachan1990** [comment]
I am getting the same error. Plain text response is working but HttpResponse.json throwing error. 

// handler 
const handlers = [
  http.get('http://localhost:3000/employees', async () => {
    return HttpResponse.json({
      message: 'hellowwww',
    })
  }),
  
  // client code
   fetch('http://localhost:3000/employees')
    .then((response) => {
      if (!response.ok) {
        throw new Error('Network response was not ok')
      }
      console.log('r====>', response.json())
      return response.statusText // This parses the response body as JSON
    })
    .then((data) => {
      console.log('====>', data)
    })
    .catch((error) => {
      console.error(
        'There has been a problem with your fetch operation:',
        error,
      )
    })
}


error:

 ====> OK
 LOG  Running "App" with {"rootTag":21,"initialProps":{}}
 LOG  [CodePush] Checking for update.
 LOG  item does not exist in storage
 LOG  [CodePush] App is up to date.
 WARN  Possible Unhandled Promise Rejection (id: 0):
SyntaxError: JSON Parse error: Unexpected end of input
SyntaxError: JSON Parse error: Unexpected end of input
    at parse (native)
    at tryCallOne (/Users/distiller/react-native/sdks/hermes/build_iphonesimulator/lib/InternalBytecode/InternalBytecode.js:53:16)

**Comment by anujsachan1990** [comment]
@willdawsonme can you please share your patch-package? 
-
Thanks

**Comment by willdawsonme** [comment]
Sure, patches below for `@msw/interceptors` and `mswjs`. You may also have to polyfill TextEncoder, TextDecoder, and [URL](https://www.npmjs.com/package/react-native-url-polyfill). URL polyfill isn't required if you're using Expo 50.

```
import { polyfillGlobal } from 'react-native/Libraries/Utilities/PolyfillFunctions';
import { TextDecoder, TextEncoder } from 'text-encoding';

polyfillGlobal('TextEncoder', () => TextEncoder);
polyfillGlobal('TextDecoder', () => TextDecoder);
```

@mswjs+interceptors+0.25.12.patch
```
diff --git a/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js b/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js
index f9dbf7e..7117967 100644
--- a/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js
+++ b/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js
@@ -58,13 +58,16 @@ var _FetchInterceptor = class extends _chunkG5ORGOKHjs.Interceptor {
       this.logger.info("awaiting for the mocked response...");
       const signal = interactiveRequest.signal;
       const requestAborted = new (0, _deferredpromise.DeferredPromise)();
-      signal.addEventListener(
-        "abort",
-        () => {
-          requestAborted.reject(signal.reason);
-        },
-        { once: true }
-      );
+      // Signal isn't always defined in react-native.
+      if (signal) {
+        signal.addEventListener(
+          "abort",
+          () => {
+            requestAborted.reject(signal.reason);
+          },
+          { once: true }
+        );
+      }
       const resolverResult = await _until.until.call(void 0, async () => {
         const listenersFinished = _chunkX3NRJIZWjs.emitAsync.call(void 0, this.emitter, "request", {
           request: interactiveRequest,
@@ -105,7 +108,9 @@ var _FetchInterceptor = class extends _chunkG5ORGOKHjs.Interceptor {
           request: interactiveRequest,
           requestId
         });
-        const response = new Response(mockedResponse.body, mockedResponse);
+        // const response = new Response(mockedResponse.body, mockedResponse);
+        // response.body is undefined in react-native.
+        const response = mockedResponse.clone();
         Object.defineProperty(response, "url", {
           writable: false,
           enumerable: true,
diff --git a/node_modules/@mswjs/interceptors/package.json b/node_modules/@mswjs/interceptors/package.json
index 3bc70f9..9cecbd3 100644
--- a/node_modules/@mswjs/interceptors/package.json
+++ b/node_modules/@mswjs/interceptors/package.json
@@ -19,7 +19,6 @@
       "default": "./lib/node/index.js"
     },
     "./ClientRequest": {
-      "browser": null,
       "types": "./lib/node/interceptors/ClientRequest/index.d.ts",
       "require": "./lib/node/interceptors/ClientRequest/index.js",
       "import": "./lib/node/interceptors/ClientRequest/index.mjs",
@@ -50,14 +49,12 @@
       "default": "./lib/node/interceptors/fetch/index.js"
     },
     "./RemoteHttpInterceptor": {
-      "browser": null,
       "types": "./lib/node/RemoteHttpInterceptor.d.ts",
       "require": "./lib/node/RemoteHttpInterceptor.js",
       "import": "./lib/node/RemoteHttpInterceptor.mjs",
       "default": "./lib/node/RemoteHttpInterceptor.js"
     },
     "./presets/node": {
-      "browser": null,
       "types": "./lib/node/presets/node.d.ts",
       "require": "./lib/node/presets/node.js",
       "import": "./lib/node/presets/node.mjs",
```

msw+2.0.6.patch
```
diff --git a/node_modules/msw/lib/native/index.js b/node_modules/msw/lib/native/index.js
index 2653060..ae8ca1f 100644
--- a/node_modules/msw/lib/native/index.js
+++ b/node_modules/msw/lib/native/index.js
@@ -43,7 +43,7 @@ __export(native_exports, {
   setupServer: () => setupServer
 });
 module.exports = __toCommonJS(native_exports);
-var import_XMLHttpRequest = require("@mswjs/interceptors/XMLHttpRequest");
+var import_fetch = require("@mswjs/interceptors/fetch");
 
 // src/node/SetupServerApi.ts
 var import_node_events = require("events");
@@ -142,7 +142,7 @@ var SetupServerApi = class extends import_SetupApi.SetupApi {
 
 // src/native/index.ts
 function setupServer(...handlers) {
-  return new SetupServerApi([import_XMLHttpRequest.XMLHttpRequestInterceptor], ...handlers);
+  return new SetupServerApi([import_fetch.FetchInterceptor], ...handlers);
 }
 // Annotate the CommonJS export names for ESM import in node:
 0 && (module.exports = {
diff --git a/node_modules/msw/package.json b/node_modules/msw/package.json
index 4239494..462a755 100644
--- a/node_modules/msw/package.json
+++ b/node_modules/msw/package.json
@@ -21,14 +21,12 @@
       "default": "./lib/browser/index.js"
     },
     "./node": {
-      "browser": null,
       "types": "./lib/node/index.d.ts",
       "require": "./lib/node/index.js",
       "import": "./lib/node/index.mjs",
       "default": "./lib/node/index.mjs"
     },
     "./native": {
-      "browser": null,
       "types": "./lib/native/index.d.ts",
       "require": "./lib/native/index.js",
       "import": "./lib/native/index.mjs",
```

**Comment by hwildwood** [comment]
Affecting me as well, I'm stuck on MSW v1 at the moment which appears to work without any patches.

**Comment by anujsachan1990** [comment]
@willdawsonme Thanks for the patch but now I have a different error after applying the suggested patch

<img width="1046" alt="Screenshot 2024-01-29 at 4 31 03 pm" src="https://github.com/mswjs/msw/assets/6928039/22647a3d-897e-4616-a4be-713f64b8a2d1">

**Comment by willdawsonme** [comment]
@anujsachan1990 Hard to say, but potentially the `if (signal) {` patch hasn't been applied. Try searching for `signal.addEventListener` in the interceptors package, there may be other codepaths that you're importing that aren't covered by my patch.

**Comment by kettanaito** [comment]
Thanks everyone for contributing with your discoveries! I've opened #2016 to provide these fixes. We also have #2016 pending to support global `fetch` in React Native. Please give those pull requests a review if you have a minute. Thank you!

**Comment by kettanaito** [comment]
## Released: v2.1.7 🎉

This has been released in v2.1.7!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v2.1.7)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/2.1.7)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.

**Comment by anujsachan1990** [comment]
@kettanaito Thanks for fixing it, it now works with fetch but still gives the same error with the rtkQuery useQuery hook. 
I am using 
import { graphqlRequestBaseQuery } from '@rtk-query/graphql-request-base-query'
instead of fetch.

**Comment by danieloi** [comment]
> Sure, patches below for `@msw/interceptors` and `mswjs`. You may also have to polyfill TextEncoder, TextDecoder, and [URL](https://www.npmjs.com/package/react-native-url-polyfill). URL polyfill isn't required if you're using Expo 50.
> 
> ```
> import { polyfillGlobal } from 'react-native/Libraries/Utilities/PolyfillFunctions';
> import { TextDecoder, TextEncoder } from 'text-encoding';
> 
> polyfillGlobal('TextEncoder', () => TextEncoder);
> polyfillGlobal('TextDecoder', () => TextDecoder);
> ```
> 
> @mswjs+interceptors+0.25.12.patch
> 
> ```
> diff --git a/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js b/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js
> index f9dbf7e..7117967 100644
> --- a/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js
> +++ b/node_modules/@mswjs/interceptors/lib/browser/chunk-5RW2GLTJ.js
> @@ -58,13 +58,16 @@ var _FetchInterceptor = class extends _chunkG5ORGOKHjs.Interceptor {
>        this.logger.info("awaiting for the mocked response...");
>        const signal = interactiveRequest.signal;
>        const requestAborted = new (0, _deferredpromise.DeferredPromise)();
> -      signal.addEventListener(
> -        "abort",
> -        () => {
> -          requestAborted.reject(signal.reason);
> -        },
> -        { once: true }
> -      );
> +      // Signal isn't always defined in react-native.
> +      if (signal) {
> +        signal.addEventListener(
> +          "abort",
> +          () => {
> +            requestAborted.reject(signal.reason);
> +          },
> +          { once: true }
> +        );
> +      }
>        const resolverResult = await _until.until.call(void 0, async () => {
>          const listenersFinished = _chunkX3NRJIZWjs.emitAsync.call(void 0, this.emitter, "request", {
>            request: interactiveRequest,
> @@ -105,7 +108,9 @@ var _FetchInterceptor = class extends _chunkG5ORGOKHjs.Interceptor {
>            request: interactiveRequest,
>            requestId
>          });
> -        const response = new Response(mockedResponse.body, mockedResponse);
> +        // const response = new Response(mockedResponse.body, mockedResponse);
> +        // response.body is undefined in react-native.
> +        const response = mockedResponse.clone();
>          Object.defineProperty(response, "url", {
>            writable: false,
>            enumerable: true,
> diff --git a/node_modules/@mswjs/interceptors/package.json b/node_modules/@mswjs/interceptors/package.json
> index 3bc70f9..9cecbd3 100644
> --- a/node_modules/@mswjs/interceptors/package.json
> +++ b/node_modules/@mswjs/interceptors/package.json
> @@ -19,7 +19,6 @@
>        "default": "./lib/node/index.js"
>      },
>      "./ClientRequest": {
> -      "browser": null,
>        "types": "./lib/node/interceptors/ClientRequest/index.d.ts",
>        "require": "./lib/node/interceptors/ClientRequest/index.js",
>        "import": "./lib/node/interceptors/ClientRequest/index.mjs",
> @@ -50,14 +49,12 @@
>        "default": "./lib/node/interceptors/fetch/index.js"
>      },
>      "./RemoteHttpInterceptor": {
> -      "browser": null,
>        "types": "./lib/node/RemoteHttpInterceptor.d.ts",
>        "require": "./lib/node/RemoteHttpInterceptor.js",
>        "import": "./lib/node/RemoteHttpInterceptor.mjs",
>        "default": "./lib/node/RemoteHttpInterceptor.js"
>      },
>      "./presets/node": {
> -      "browser": null,
>        "types": "./lib/node/presets/node.d.ts",
>        "require": "./lib/node/presets/node.js",
>        "import": "./lib/node/presets/node.mjs",
> ```
> 
> msw+2.0.6.patch
> 
> ```
> diff --git a/node_modules/msw/lib/native/index.js b/node_modules/msw/lib/native/index.js
> index 2653060..ae8ca1f 100644
> --- a/node_modules/msw/lib/native/index.js
> +++ b/node_modules/msw/lib/native/index.js
> @@ -43,7 +43,7 @@ __export(native_exports, {
>    setupServer: () => setupServer
>  });
>  module.exports = __toCommonJS(native_exports);
> -var import_XMLHttpRequest = require("@mswjs/interceptors/XMLHttpRequest");
> +var import_fetch = require("@mswjs/interceptors/fetch");
>  
>  // src/node/SetupServerApi.ts
>  var import_node_events = require("events");
> @@ -142,7 +142,7 @@ var SetupServerApi = class extends import_SetupApi.SetupApi {
>  
>  // src/native/index.ts
>  function setupServer(...handlers) {
> -  return new SetupServerApi([import_XMLHttpRequest.XMLHttpRequestInterceptor], ...handlers);
> +  return new SetupServerApi([import_fetch.FetchInterceptor], ...handlers);
>  }
>  // Annotate the CommonJS export names for ESM import in node:
>  0 && (module.exports = {
> diff --git a/node_modules/msw/package.json b/node_modules/msw/package.json
> index 4239494..462a755 100644
> --- a/node_modules/msw/package.json
> +++ b/node_modules/msw/package.json
> @@ -21,14 +21,12 @@
>        "default": "./lib/browser/index.js"
>      },
>      "./node": {
> -      "browser": null,
>        "types": "./lib/node/index.d.ts",
>        "require": "./lib/node/index.js",
>        "import": "./lib/node/index.mjs",
>        "default": "./lib/node/index.mjs"
>      },
>      "./native": {
> -      "browser": null,
>        "types": "./lib/native/index.d.ts",
>        "require": "./lib/native/index.js",
>        "import": "./lib/native/index.mjs",
> ```

This cleared most of my react native issues. Thanks so much 😭
Creating my own patch for current versions wasn't too dissimilar from what you shared.
If you're using axios, you want to skip the patch for msw and keep the one for the interceptor. 
That's because axios uses XMLHTTP... not fetch.

**Comment by danieloi** [comment]
@kettanaito I'm using axios in my project and it uses xmlHttpRequest not the fetch API. 
I'm still getting empty bodies in my mocked responses with axios. Fetch works fine

**Comment by rohan-sr** [comment]
> @kettanaito Thanks for fixing it, it now works with fetch but still gives the same error with the rtkQuery useQuery hook. I am using import { graphqlRequestBaseQuery } from '@rtk-query/graphql-request-base-query' instead of fetch.

@anujsachan1990 Any luck on this ? I am also using rtkQuery and useQuery hook along with 
`import { graphqlRequestBaseQuery } from '@rtk-query/graphql-request-base-query',` 
requests are intercepted in resolver function , but mocked responses are not received

**Comment by danieloi** [comment]
> still getting empty bodies in my moc

solved this by using version: ^1.3.2
